celery/utils/functional.py

```diff
@@ -182,38 +182,73 @@ def __init__(self, it):
         self.__it = it
         self.__index = 0
         self.__consumed = []
+        self.__done = False
 
     def __reduce__(self):
         return list, (self.data,)
 
     def __length_hint__(self):
         return self.__it.__length_hint__()
 
+    def __repr__(self):
+        # override list.__repr__ to avoid consuming the generator
+        if self.__done:
+            return repr(self.__consumed)
+        else:
+            return '[{0}]'.format(', '.join(
+                [repr(x) for x in self.__consumed] + ['...'])
+            )
+
     def __iter__(self):
-        return chain(self.__consumed, self.__it)
+        for x in self.__consumed:
```

Comment:
`yield from ` is shorter...

Reply:
Yep, I'll change this and have a more critical read through for other 2.7isms which could be removed from this diff.